include/aws/crt/Variant.h

```diff
@@ -0,0 +1,595 @@
+                AWS_ASSERT(m_index != -1);
+            }
+
+            Variant &operator=(const Variant &other)
+            {
+                if (this != &other)
+                {
+                    if (m_index != other.m_index)
+                    {
+                        Destroy();
+                        CopyConstructUtil<0, Ts...>(other);
+                    }
+                    else
+                    {
+                        CopyAssignUtil<0, Ts...>(other);
+                    }
+                }
+                return *this;
+            }
+
+            Variant &operator=(Variant &&other)
+            {
+                if (this != &other)
+                {
+                    if (m_index != other.m_index)
+                    {
+                        Destroy();
+                        MoveConstructUtil<0, Ts...>(std::move(other));
+                    }
+                    else
+                    {
+                        MoveAssignUtil<0, Ts...>(std::move(other));
+                    };
+                }
+                return *this;
+            }
+
+            /* emplace */
+            template <typename T, typename... Args, EnableIfOtherIsThisVariantAlternative<T> = 1>
+            T &emplace(Args &&...args)
+            {
+                static_assert(
+                    VariantDetail::Checker::HasType<typename std::decay<T>::type, Ts...>::value,
+                    "This variant does not have such alternative T.");
+                static_assert(
+                    sizeof(T) <= STORAGE_SIZE,
+                    "Attempting to instantiate a Variant with a type bigger than all alternatives.");
+
+                Destroy();
+
+                using PlainT = typename std::decay<T>::type;
+                new (m_storage) PlainT(std::forward<Args>(args)...);
+                m_index = VariantDetail::Index::GetIndexOf<T, Ts...>();
+                AWS_ASSERT(m_index != -1);
+
+                T *value = reinterpret_cast<T *>(m_storage);
+                return *value;
+            }
+
+            template <std::size_t Index, typename... Args>
+            auto emplace(Args &&...args) -> typename ThisVariantAlternative<Index>::type &
+            {
+                static_assert(Index < AlternativeCount, "Unknown alternative index to emplace");
+                using AlternativeT = typename ThisVariantAlternative<Index>::type;
+
+                return emplace<AlternativeT, Args...>(args...);
+            }
+
+            template <typename T, EnableIfOtherIsThisVariantAlternative<T> = 1> bool holds_alternative() const
+            {
+                AWS_ASSERT(m_index != -1);
+                return m_index == VariantDetail::Index::GetIndexOf<T, Ts...>();
+            }
+
+            template <size_t Index> constexpr bool holds_alternative() const { return Index < AlternativeCount; }
+
+            template <typename T> bool holds_alternative() const { return false; }
```

Comment:
I am not a master of template programming. Should this be: `template <typename T, EnableIfOtherIsThisVariantAlternative<T> = 0> bool holds_alternative() const`???  I see this is the fallback to the other `holds_alternative()` that uses `enable_if`. Is this how you're supposed to implement a fallback? Or is it supposed to be more explicit?

But also, [docs](https://en.cppreference.com/w/cpp/utility/variant/holds_alternative) say:
> The call is ill-formed if T does not appear exactly once in Types...

so maybe we shouldn't even implement this?

Reply:
A good catch!
I've moved it to private cause it is pretty handy.
Also, I've removed generic version for types not present in Variant (as it does not exist for std variant).